oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/backend/MapReducerIgniteLocalPeek.java

```diff
@@ -225,33 +228,30 @@ public boolean isActive() {
 
     public abstract S execute(Ignite node);
 
-    List<Pair<IgniteCache<Long, GridOSHEntity>, Long>> localKeys(Ignite node) {
-      // calculate all cache keys we have to investigate
-      List<Pair<IgniteCache<Long, GridOSHEntity>, Long>> localKeys = new ArrayList<>();
-      this.cacheNames.forEach(cacheName -> {
+    Stream<Pair<IgniteCache<Long, GridOSHEntity>, Long>> localKeys(Ignite node) {
+      return this.cacheNames.stream().flatMap(cacheName -> {
         IgniteCache<Long, GridOSHEntity> cache = node.cache(cacheName);
-        List<Long> cellIdRangeIds = new ArrayList<>();
-        this.cellIdRanges.forEach(cellIdRange -> {
-          cellIdRangeIds.clear();
-          int level = cellIdRange.getLeft().getZoomLevel();
-          long from = CellId.getLevelId(level, cellIdRange.getLeft().getId());
-          long to = CellId.getLevelId(level, cellIdRange.getRight().getId());
-          for (long key = from; key <= to; key++) {
-            cellIdRangeIds.add(key);
-          }
-          // Map keys to ignite nodes and remember the local ones
-          node.<Long>affinity(cache.getName())
-              .mapKeysToNodes(cellIdRangeIds)
-              .getOrDefault(node.cluster().localNode(), Collections.emptyList())
-              .forEach(key -> localKeys.add(new ImmutablePair<>(cache, key)));
-        });
+        return StreamSupport.stream(this.cellIdRanges.spliterator(), false)
+            .flatMap(cellIdRange -> {
+              int level = cellIdRange.getLeft().getZoomLevel();
+              long fromId = cellIdRange.getLeft().getId();
+              long toId = cellIdRange.getRight().getId();
+              List<Long> cellIdRangeKeys = LongStream.rangeClosed(
+                  CellId.getLevelId(level, fromId),
+                  CellId.getLevelId(level, toId)
+              ).boxed().collect(Collectors.toList());
+              // Map keys to ignite nodes and remember the local ones
+              return node.<Long>affinity(cache.getName())
+                  .mapKeysToNodes(cellIdRangeKeys)
+                  .getOrDefault(node.cluster().localNode(), Collections.emptyList())
+                  .stream()
+                  .map(key -> new ImmutablePair<>(cache, key));
+            });
       });
-      Collections.shuffle(localKeys);
-      return localKeys;
     }
 
     S execute(Ignite node, CellProcessor<S> cellProcessor) {
-      return this.localKeys(node).parallelStream()
+      return this.localKeys(node).parallel()
```

Comment:
I doubt that .parallel() has any effect at this point, cellIdRange in its current implementation is not parallelizable, and so the rest of the stream.
Suggestion is to get an Iterator  `this.localKeys(node).iterator()` then batch up some localKeys and reduce them in parallel.
```java
ArrayList<Object> batch = new ArrayList<>(parallelism);
while(itr.hasNext()){
	batch.add(itr.next());
	if(batch.size() == parallelism){
		batch.stream()
		.parallel();
		// map&filter
		// reduce batch
		batch.clear();
	}
}
batch.stream()
.parallel();
// map&filter
// reduce batch
batch.clear();
//reduce batch results together
```

Reply:
> I doubt that .parallel() has any effect at this point

Are you sure? If I try something like this, it seems like `.parallel` can be called on a stream generated from `StreamSupport.stream(…, false)`:

```java
Iterable<Integer> foo = () -> (new ArrayList<Integer>()).iterator();
System.out.println(StreamSupport.stream(foo.spliterator(), false).isParallel());
System.out.println(StreamSupport.stream(foo.spliterator(), false).parallel().isParallel());
```

This returns `false`, then `true` (`isParallel` returns whether the stream, […] would execute in parallel).

:thinking:

Reply:
```java
final int limit = 1_000;

Iterable<Long> test = () -> new Iterator<Long>(){
	private long l = 0;
	@Override
	public boolean hasNext() {
		return l < limit;
	}

	@Override
	public Long next() {
		if(!hasNext())
			throw new NoSuchElementException();
		return Long.valueOf(l++);
	}
};

Consumer<Long> burn10ms = it -> {
	try {
		Thread.sleep(10);
	} catch (InterruptedException e) {
	}
};

Stopwatch stopwatch = Stopwatch.createStarted();
StreamSupport.stream(test.spliterator(), true).parallel()
.forEach(burn10ms);
System.out.println(""+stopwatch);

ArrayList<Long> test2 = new ArrayList<>(limit);
for(int i=0; i< limit; i++){
	test2.add(Long.valueOf(i));
}
stopwatch.reset().start();
StreamSupport.stream(test2.spliterator(), false).parallel()
.forEach(burn10ms);
System.out.println(""+stopwatch);
```
According to this test, it would not be parallelized.

Reply:
> I'm not sure if isParallel return if the stream is really parallel.

According to the javadoc it should: https://docs.oracle.com/javase/8/docs/api/java/util/stream/BaseStream.html#isParallel--

> According to this test

When I try it with a stream with more than only 1,000 elements, it seems to run in parallel just fine (I replaced the `Thread.sleep(10)` with a loop that burs a "few" CPU cycles, and can get all my CPUs working at close to 100%).